### PR TITLE
MVV - Minimal Viable Vagrantfile

### LINKED ISSUE #546: Try recommending a Vagrantfile
@jalessio showed a basic Vagrant setup in https://gist.github.com/jalessio/452ec1e7f75b222867f0ccae36e621ba, which might be useful for easier local development without requiring changes to the Chef setup. Experiment with this to see if it improves documentation and approachability of code.

### PR DESCRIPTION
**Pull request by jalessio**
Supersedes PR #548 as a branch in this repo instead of in [my fork](https://github.com/jalessio/machine/tree/add-vagrant). 

This was rebased off `migurski/fix-la-reunion-encoding` so it should include the "La Réunion" fix in 681373e6f and in turn allow all tests to pass (I'm open to being told that was a bad idea).

To do:
- [x] Add a working Vangrantfile
- [x] Verify that unit tests pass in Virtualbox with shared filesystem
- [x] Update installation docs to reflect Vagrant usage
- [x] Wait for test feedback

Closes #546.

### CODE REVIEW

**Comment by migurski** [comment]
This is great! if you don't mind I'll make some modifications to the documentation to reflect use of the Vagrantfile, and then we can merge. I'm adding some checks to the PR description.

**Comment by jalessio** [comment]
Yep -- have at it and make whatever changes you want. I'm building it now locally to verify tests pass.

**Comment by migurski** [comment]
All right, docs updated! I've spoke with @trescube about being a guinea pig for this work, because he graciously suffered through the previous manual VirtualBox version.

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
+
+Vagrant.configure(2) do |config|
+  config.vm.box     = "ubuntu/trusty64"
```
Comment (jalessio) [inline Vagrantfile]: unnecessary whitespace?

File: docs/install.md [hunk 2]
```diff
@@ -1,38 +1,69 @@
...
-everything onto an Ubuntu Linux virtual machine. We recommend using a dedicated
-virtual machine for development; our Chef script installs servers, packages, and
-other configuration that may disrupt other uses of an Ubuntu machine. This
-process should take about 20 minutes depending on download speed, or you can
-skip the first four steps by using a pre-built Ubuntu 14.04 image on Amazon EC2.
+everything onto a [Vagrant virtual machine](https://www.vagrantup.com). We
+recommend using a dedicated virtual machine for development; our Chef script
+installs servers, packages, and other configuration that may disrupt other uses
+of an Ubuntu machine. This process should take about 10-20 minutes depending on
+download speed.
```
Comment (jalessio) [inline docs/install.md]: Whattaya think about this instead so we don't confuse folks out of the dates regarding the relationship between Vagrant and VirtualBox?

> You can edit a local copy of OpenAddresses code with working tests by installing everything onto a local virtual machine using [VirtualBox](https://www.virtualbox.org/) and [Vagrant](https://www.vagrantup.com).

**Comment by trescube** [comment]
I'm running on Mac OSX 10.12.3. 

I followed the [instructions](https://github.com/openaddresses/machine/blob/a473b918032db1c70237b9b053a774bff61b7d1c/docs/install.md).  VirtualBox was already installed at version 5.0.32 r112930.  Vagrant was not installed so I did so and ended up with Vagrant 1.9.1.  `VBoxManage` is in my path.  I cloned the [machine](https://github.com/openaddresses/machine) and ran `vagrant up` from within the machine directory and got:

```
[stephenhess@slumlord machine (master)]$ vagrant up
Vagrant failed to initialize at a very early stage:

The plugins failed to initialize correctly. This may be due to manual
modifications made within the Vagrant home directory. Vagrant can
attempt to automatically correct this issue by running:

  vagrant plugin repair

If Vagrant was recently updated, this error may be due to incompatible
versions of dependencies. To fix this problem please remove and re-install
all plugins. Vagrant can attempt to do this automatically by running:

  vagrant plugin expunge --reinstall

Error message given during initialization: Unable to resolve dependency: user requested 'vagrant-omnibus (> 0)'
```

I took that advice and ran:

```
[stephenhess@slumlord machine (master)]$ vagrant plugin repair
Repairing currently installed plugins. This may take a few minutes...
Fetching: vagrant-omnibus-1.5.0.gem (100%)
Fetching: vagrant-berkshelf-5.1.1.gem (100%)
The Vagrant Berkshelf plugin requires Berkshelf from the Chef Development Kit.
You can download the latest version of the Chef Development Kit from:

    https://downloads.chef.io/chef-dk/

Installing Berkshelf via other methods is not officially supported.
Installed plugins successfully repaired!
```

I tried running `vagrant up` again and got:

```
[stephenhess@slumlord machine (master)]$ vagrant up
A Vagrant environment or target machine is required to run this
command. Run `vagrant init` to create a new Vagrant environment. Or,
get an ID of a target machine from `vagrant global-status` to run
this command on. A final option is to change to a directory with a
Vagrantfile and to try again.
```

I ran `vagrant init` and got:

```
[stephenhess@slumlord machine (master)]$ vagrant init
A `Vagrantfile` has been placed in this directory. You are now
ready to `vagrant up` your first virtual environment! Please read
the comments in the Vagrantfile as well as documentation on
`vagrantup.com` for more information on using Vagrant.
```

That seemed like a good thing so I ran `vagrant up` again and got:

```
[stephenhess@slumlord machine (master)]$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Box 'base' could not be found. Attempting to find and install...
    default: Box Provider: virtualbox
    default: Box Version: >= 0
==> default: Box file was not detected as metadata. Adding it directly...
==> default: Adding box 'base' (v0) for provider: virtualbox
    default: Downloading: base
An error occurred while downloading the remote file. The error
message, if any, is reproduced below. Please fix this error and try
again.

Couldn't open file /Users/stephenhess/git/trescube/machine/base
```

That's where I'm stuck.

File: docs/install.md (hunk 2 above)
Comment (trescube) [inline docs/install.md]: I was confused at first about whether I needed to start a VirtualBox Ubuntu instance in preparation for using Vagrant.  It's my understanding that Vagrant initializes a VirtualBox ubuntu instance.  Is that correct?

**Comment by trescube** [comment]
*Disclaimer* - I know very little about ops work and technologies so don't assume that I know anything other than how to follow instructions.

**Comment by migurski** [comment]
Thanks Stephen, this is really useful! I ran into some weird gotchas with Vagrant as well because I did not install via Homebrew. I'll try to figure out why you encountered these errors.

File: docs/install.md (hunk 2 above)
Comment (jalessio) [inline docs/install.md]: Correct, Vagrant reads the `Vagrantfile` to see what VM to download, downloads it, and then hands it off to VirtualBox where it starts running. VirtualBox itself should be mostly invisible.

**Comment by jalessio** [comment]
@trescube that's really strange -- it kinda looks like you had a previous install of Vagrant on that machine? Is that possible? Does the output of `ls -l ~/.vagrant.d` show files that were created before today?

I suggest you move your `~/.vagrant.d` directory out of the way and try again (something like `mv ~/.vagrant.d /.vagrant.d.bak`) then give `vagrant up` another try from where you cloned the `machine` repo.

**Comment by migurski** [comment]
I'm confused in particular about why it wants these Chef-related plugins like “vagrant-omnibus”; I wonder it it’s trying to infer something from the presence of the “chef” directory? Vagrant should know nothing of Chef for this project.

**Comment by jalessio** [comment]
@trescube Also, from your comments it looks like you were trying to run this from the master branch?
```
[stephenhess@slumlord machine (master)]$ vagrant up
```
That would most likely explain why the `Vagrantfile` didn't exist, right? (since it's only on the `jalessio/add-vagrant` branch so far)

**Comment by migurski** [comment]
That would potentially explain why it’s looking for “base”. My other thought is that there are some leftovers [from `pelias/vagrant`](https://github.com/pelias/vagrant/blob/master/Vagrantfile), another project that @trescube is likely to have used Vagrant for in the past.

**Comment by jalessio** [comment]
Yeah, I just ran `vagrant init` to see what's in the `Vagrantfile` it generates and that explains "base":
```
config.vm.box = "base"
```

**Comment by trescube** [comment]
@jalessio Ah, I think you're right, I probably have an older vagrant sitting around.  I'll follow your instructions and try again today.

**Comment by trescube** [comment]
```
[stephenhess@slumlord machine (master)]$ ll ~/.vagrant.d
total 24
drwxr-xr-x   6 stephenhess  staff   204 Jan 22  2016 boxes
drwxr-xr-x   9 stephenhess  staff   306 Feb 22 23:16 data
drwxr-xr-x  11 stephenhess  staff   374 Feb 22 23:07 gems
-rw-------   1 stephenhess  staff  1675 Dec 21  2015 insecure_private_key
-rw-r--r--   1 stephenhess  staff   355 Jan 22  2016 plugins.json
drwxr-xr-x   3 stephenhess  staff   102 Dec 21  2015 rgloader
-rw-r--r--   1 stephenhess  staff     3 Dec 21  2015 setup_version
drwxr-xr-x   2 stephenhess  staff    68 Feb 22 23:18 tmp
```

**Comment by trescube** [comment]
I did a clean clone of machine `master`, then `co`'d jalessio/add-vagrant, followed by a `vagrant up` and it's humming along nicely now.

**Comment by trescube** [comment]
Okay, it finished and `vagrant ssh` logs me into the VM.  Dumb question, though, when I `vagrant ssh` into the VM, I'll have to git clone the machine and operate with it from there, correct?  That is, there's no pass-through between the local machine directory and the VM?

**Comment by migurski** [comment]
You won't need to do anything else, and you should just run the tests. The `/vagrant` directory in the virtual machine is the same as your current host directory. Try touching files to see that they change in both.

**Comment by migurski** [comment]
I added a note about the `/vagrant` mount to the install docs. I think this will be good to merge as soon as the tests pass.

**Comment by migurski** [comment]
Okay, merging. Thank you @trescube and @jalessio, this should make the code much more approachable for people!

**Comment by jalessio** [comment]
@migurski might be less confusing if we mount the code into `/home/vagrant/openaddresses/machine` or just `/home/vagrant/machine`? That way after `vagrant ssh` there's something obvious in your current working directory. It's a one-liner addition to the `Vagrantfile` to make that happen.

**Comment by trescube** [comment]
@migurski Yes, much more approachable, and I can work directly within atom again!  

@jalessio Mounting in `/home/vagrant/machine` would be less confusing, I was thrown off by the empty directory when I first `vagrant ssh`'d

**Comment by migurski** [comment]
Yes, I agree. I didn't realize that was possible.

**Comment by jalessio** [comment]
PR for this on its way.
